src/libsyntax/parse/mod.rs

```diff
@@ -286,52 +286,49 @@ pub fn tts_to_parser<'a>(sess: &'a ParseSess,
 pub fn char_lit(lit: &str) -> (char, isize) {
     use std::char;
 
-    let mut chars = lit.chars();
-    match (chars.next(), chars.next()) {
-        (Some(c), None) if c != '\\' => return (c, 1),
-        (Some('\\'), Some(c)) => match c {
-            '"' => return ('"', 2),
-            'n' => return ('\n', 2),
-            'r' => return ('\r', 2),
-            't' => return ('\t', 2),
-            '\\' => return ('\\', 2),
-            '\'' => return ('\'', 2),
-            '0' => return ('\0', 2),
-            _ => {}
-        },
-        _ => panic!("lexer accepted invalid char escape `{}`", lit)
-    };
-
-    fn esc(len: usize, lit: &str) -> Option<(char, isize)> {
-        u32::from_str_radix(&lit[2..len], 16).ok()
-        .and_then(char::from_u32)
-        .map(|x| (x, len as isize))
+    let err = |i| format!("lexer accepted invalid char literal {} step {}", lit, i);
+
+    if lit.as_bytes()[0] != b'\\' {
+        // If the first byte isn't '\\' it might part of a multi-byte char, so
+        // get the char with chars().
+        let c = lit.chars().next().unwrap_or_else(|| {
+            panic!(err(0));
+        });
+        return (c, 1);
     }
 
-    let unicode_escape = || -> Option<(char, isize)> {
-        if lit.as_bytes()[2] == b'{' {
+    match lit.as_bytes()[1] as char {
```

Comment:
> BTW, do these comments constitute a proper review, or are they just "drive-by" comments? Will the commit receive attention from anybody else?

They do constitute a proper review. Usually a bot would have commented and assigned a reviewer for you, but seems like it has bugged out for some reason.

> It's not clear to me what I need to do now for this PR to receive r+.

Now that I read it again, it sort of makes sense to handle no-escape case outside of escape case. 

Perhaps something like following could be better:

```rust
if first_byte == '\\' { 
    /* handle escape */ 
} else { 
    /* handle non-escape case */ 
}
```

so the escape handling is still tied to the `\\` that is supposed to go before the escape sequence, rather than to the no-escape case like it is now.

Once you adjust the PR for error handling and making `\\` thing more tied to parsing of escape sequences (in any way), I feel this PR would be in a great shape to r+.